### DISCUSSION TITLE
【求助】如何使用自定义的Editor替换内置CodeEditor

### DISCUSSION THREAD

**Discussion by MilkWangStudio** [opensumi/core#3013] [Q&A]
目前Editor的Tool面板是在右上角，我需要修改下，将操作栏放到Tab栏下方，我看CodeEditor的组件里这部分没办法配置，所以我想自己实现一个Editor，但是又想继承对于和FileTree的联动、自动保存等特性，请问有什么思路吗~

<img width="1788" alt="image" src="https://github.com/opensumi/core/assets/15866249/53d3ca74-0902-473b-b0dd-0e0a290c52b6">

参考
![image](https://github.com/opensumi/core/assets/15866249/443523d4-950c-46cc-b7d1-15d574fd6c7b)

**Comment by erha19** [comment]
@MilkWangStudio 目前底部可以使用 `BrowserEditorContribution` 贡献点注册自定义组件进行替换，伪代码如下：

```ts
import {
  BrowserEditorContribution,
  EditorComponentRegistry,
} from '@opensumi/ide-editor/lib/browser';

@Domain(BrowserEditorContribution)
export class CustomNavigationContribution implements BrowserEditorContribution {
...

registry.registerEditorSideWidget({
      id: CUSTOM_NAVIGATION_WIDGET_ID,
      component: CustomNavigation,
      // 只有 git 协议的文件下才显示
      displaysOnResource: (r) => {
        return r.uri.scheme === 'git';
      },
});
...
}
```
可以参考这块逻辑提个 PR 适配一下顶部场景，代码部分应该不会很复杂

**Reply by MilkWangStudio** [reply]
好的，感谢回复，我试一下，后续项目开发完我把改的一些代码抽出来提一些PR~
